File: java/poker/src/main/java/com/duchen/poker/model/type/TypeHelper.java
package com.duchen.poker.model.type;

import com.duchen.poker.model.CardGroup;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TypeHelper {

    private static TypeHelper sInstance;

    public static TypeHelper getInstance() {
        if (sInstance == null) {
            sInstance = new TypeHelper();
        }
        return sInstance;
    }

    List<TypeChecker> mCheckers = new ArrayList<>();

    private TypeHelper() {
        mCheckers.add(new ZeroTypeChecker());
        mCheckers.add(new SingleTypeChecker());
        mCheckers.add(new DoubleTypeChecker());
        mCheckers.add(new ThreeTypeChecker());
        mCheckers.add(new SingelLineTypeChecker());
        mCheckers.add(new DoublelLineTypeChecker());
        mCheckers.add(new ThreeLineTypeChecker());
        mCheckers.add(new ThreeTakeOneTypeChecker());
        mCheckers.add(new ThreeTakeTwoTypeChecker());
        mCheckers.add(new ThreeTakeOneLineTypeChecker());
        mCheckers.add(new ThreeTakeTwoLineTypeChecker());
        mCheckers.add(new BombTypeChecker());
        mCheckers.add(new FourTakeOneTypeChecker());
        mCheckers.add(new FourTakeTwoTypeChecker());
        mCheckers.add(new KingTypeChecker());
    }

    public CardGroup isOneCardGroup(List<Integer> cards) {
        int count = cards.size();
        for (TypeChecker checker : mCheckers) {
            if (checker.getMyPossibleCardCount().contains(count)) {
                int maxCard = checker.getMaxCardIfIsMyType(cards);
                if (maxCard >= 0) {
                    CardGroup group = new CardGroup();
                    if (!cards.isEmpty()) {
                        Integer[] cardNums = (Integer[]) cards.toArray();
                        Arrays.sort(cardNums);
                        group.setCardList(Arrays.asList(cardNums));
                    }
                    group.setCardGroupType(checker.getType());
                    group.setMaxCard(maxCard);
                    return group;
                }
            }
        }
        CardGroup error = new CardGroup();
        error.setCardGroupType(CardGroupType.ERROR);
        Integer[] cardNums = (Integer[]) cards.toArray();
        Arrays.sort(cardNums);
        error.setCardList(Arrays.asList(cardNums));
        return error;
    }
}

File: java/poker/src/main/kotlin/poker/model/type/TypeHelper.kt
package com.duchen.poker.model.type

import com.duchen.poker.model.CardGroup

import java.util.ArrayList
import java.util.Arrays

class TypeHelper private constructor() {

    internal var mCheckers: MutableList<TypeChecker> = ArrayList()

    init {
        mCheckers.add(ZeroTypeChecker())
        mCheckers.add(SingleTypeChecker())
        mCheckers.add(DoubleTypeChecker())
        mCheckers.add(ThreeTypeChecker())
        mCheckers.add(SingleLineTypeChecker())
        mCheckers.add(DoubleLineTypeChecker())
        mCheckers.add(ThreeLineTypeChecker())
        mCheckers.add(ThreeTakeOneTypeChecker())
        mCheckers.add(ThreeTakeTwoTypeChecker())
        mCheckers.add(ThreeTakeOneLineTypeChecker())
        mCheckers.add(ThreeTakeTwoLineTypeChecker())
        mCheckers.add(BombTypeChecker())
        mCheckers.add(FourTakeOneTypeChecker())
        mCheckers.add(FourTakeTwoTypeChecker())
        mCheckers.add(KingTypeChecker())
    }

    fun isOneCardGroup(cards: MutableList<Int>): CardGroup {
        val count = cards.size
        for (checker in mCheckers) {
            if (checker.myPossibleCardCount.contains(count)) {
                val maxCard = checker.getMaxCardIfIsMyType(cards)
                if (maxCard >= 0) {
                    val group = CardGroup()
                    group.cardList = cards.sorted()
                    group.cardGroupType = checker.type
                    group.maxCard = maxCard
                    return group
                }
            }
        }
        val error = CardGroup()
        error.cardGroupType = CardGroupType.ERROR
        error.cardList = cards.sorted()
        return error
    }

    companion object {
        var instance = TypeHelper()
    }
}
